TsRandomizer/Randomisation/OrbExperienceManager.cs

```diff
@@ -0,0 +1,105 @@
+		private static readonly Type LunaisOrb = TimeSpinnerType.Get("Timespinner.GameObjects.Heroes.Orbs.LunaisOrb");
+		private static readonly Type LunaisSpell = TimeSpinnerType.Get("Timespinner.GameObjects.Heroes.Spells.LunaisSpell");
+		private static readonly Type LunaisOrbAbility = TimeSpinnerType.Get("Timespinner.GameObjects.Heroes.Lunais.BaseClasses.LunaisOrbAbility");
+		private static readonly Type LunaisSpellManager = TimeSpinnerType.Get("Timespinner.GameObjects.Heroes.Spells.LunaisSpellManager");
+		private static readonly PropertyInfo mainOrbProperty = LunaisOrbManager.GetProperty("MainOrb", BindingFlags.Instance | BindingFlags.Public | BindingFlags.DeclaredOnly);
+		private static readonly PropertyInfo subOrbProperty = LunaisOrbManager.GetProperty("SubOrb", BindingFlags.Instance | BindingFlags.Public | BindingFlags.DeclaredOnly);
+		private static readonly PropertyInfo spellProperty = LunaisSpellManager.GetProperty("EquippedSpell", BindingFlags.Instance | BindingFlags.Public | BindingFlags.DeclaredOnly);
+		private static readonly PropertyInfo orbColorProperty = LunaisOrb.GetProperty("OrbColor", BindingFlags.Instance | BindingFlags.Public | BindingFlags.DeclaredOnly);
+		private static readonly PropertyInfo spellTypeProperty = LunaisSpell.GetProperty("SpellType", BindingFlags.Instance | BindingFlags.Public | BindingFlags.DeclaredOnly);
+
+		public static void UpdateHitRegistry(Protagonist lunais)
+		{
+			var hitEnemyRegistryProperty = LunaisOrbAbility.GetProperty("HitEnemyRegistry", BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.DeclaredOnly);
+			var orbManager = lunais.AsDynamic()._orbManager;
+			var spellManager = lunais.AsDynamic()._spellManager;
+			var mainOrb = mainOrbProperty.GetValue(orbManager, null);
+			if (mainOrb != null)
+				MainOrbDamagedEnemies.UnionWith(hitEnemyRegistryProperty.GetValue(mainOrb, null));
+
+			var subOrb = subOrbProperty.GetValue(orbManager, null);
+			if (subOrb != null)
+				SubOrbDamagedEnemies.UnionWith(hitEnemyRegistryProperty.GetValue(subOrb, null));
+
+			var spell = spellProperty.GetValue(spellManager, null);
+			if (spell != null)
+				SpellDamagedEnemies.UnionWith(hitEnemyRegistryProperty.GetValue(spell, null));
+		}
+
+		public static void ResetHitRegistry()
+		{
+			MainOrbDamagedEnemies.Clear();
+			SubOrbDamagedEnemies.Clear();
+			SpellDamagedEnemies.Clear();
+		}
+
+		public static void UpdateOrbXp(Level level, Protagonist lunais, double extraXp)
+		{
+			var inventory = level.GameSave.Inventory;
+			var hasEarringsEquipped = inventory.EquippedTrinketA == EInventoryEquipmentType.NelisteEarring
+					|| inventory.EquippedTrinketB == EInventoryEquipmentType.NelisteEarring;
+
+			if (!hasEarringsEquipped) return;
+			// minus the xp from the base game, minus the xp from the call to GiveOrbExperience to run the level up logic
+			int extraXpToAdd = (int)(extraXp - 2);
+			var levelReflected = level.AsDynamic();
+			dynamic orbManager = lunais.AsDynamic()._orbManager;
```

Comment:
plz dont do multile calls to `lunais.AsDynamic()` instead make it variable

Reply:
One of these is level, not lunais. Although i can pass in the exact property i need from level reflected from levelobject.cs so I'll do that

Reply:
Oh. Github on my phone didn't show me the proper code. Yeah I'll fix this

Reply:
i have done the fixing